cpp/ql/src/Security/CWE/CWE-078/ExecTainted.ql

```diff
@@ -15,63 +15,113 @@
 import cpp
 import semmle.code.cpp.security.CommandExecution
 import semmle.code.cpp.security.Security
-import DataFlow::PathGraph
 import semmle.code.cpp.ir.dataflow.TaintTracking
+import semmle.code.cpp.ir.dataflow.TaintTracking2
 import semmle.code.cpp.ir.IR
 import semmle.code.cpp.security.FlowSources
+import semmle.code.cpp.models.implementations.Strcat
 
 Expr sinkAsArgumentIndirection(DataFlow::Node sink) {
-  result = sink.asOperand()
-      .(SideEffectOperand)
-      .getAddressOperand()
-      .getAnyDef()
-      .getUnconvertedResultExpression()
+  result =
+    sink.asOperand()
+        .(SideEffectOperand)
+        .getAddressOperand()
+        .getAnyDef()
+        .getUnconvertedResultExpression()
 }
 
-class ExecTaintConfiguration extends TaintTracking::Configuration {
-  ExecTaintConfiguration() { this = "ExecTaintConfiguration" }
+predicate interestingConcatenation(DataFlow::Node fst, DataFlow::Node snd) {
+  exists(FormattingFunctionCall call, int index, FormatLiteral literal |
+    sinkAsArgumentIndirection(fst) = call.getConversionArgument(index) and
+    snd.asDefiningArgument() = call.getOutputArgument(false) and
+    literal = call.getFormat() and
+    not literal.getConvSpecOffset(index) = 0 and
+    (
+      literal.getConversionType(index) instanceof CharPointerType
+      or
+      literal.getConversionType(index).(PointerType).getBaseType() instanceof Wchar_t
```

Comment:
You will also occasionally see `char16_t *` strings, e.g. in ChakraCore.  Could we widen this to any pointer to an `IntegralType` (or perhaps look at the `getConversionChar` instead)?

Reply:
Good point - I think `getConversionChar` is the right way to go.